Add support for NOAA tide information

jcconnell opened the pull request:
## Description:

This sensor adds support for NOAA Tide information.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5881

## Example entry for `configuration.yaml` (if applicable):
```yaml
senor:
  - platform: noaa_tides
    station_id: 8721164
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New files were added to `.coveragerc`.

awarecan committed:
Frontend component should auto load auth coomponent (#15606)

cdce8p committed:
Bugfix HomeKit name and serial_number (#15600)

* Bugfix HomeKit name and serial_number

* Revert serial_number changes

amelchio committed:
Use case insensitive comparison for Sonos model check (#15604)

balloob committed:
Cast/Sonos: create config entry if manually configured (#15630)

* Cast/Sonos: create config entry if manually configured

* Add test for helper

balloob committed:
Bumped version to 0.74.1

balloob committed:
Merge pull request #15651 from home-assistant/rc

0.74.1

jcconnell committed:
Add support for NOAA tide information

jcconnell committed:
Fixing indentation line 120 flake8 vs pylint

jcconnell commented:
It looks like flake8 and pylint have conflicting goals for the indentation of line 120.

balloob committed:
Fix CORS duplicate registration (#15670)

balloob committed:
Bumped version to 0.74.2

balloob committed:
0.74.2 (#15671)

* Fix CORS duplicate registration (#15670)

* Bumped version to 0.74.2

MartinHjelmare commented on .coveragerc:
```diff
@@ -667,6 +667,7 @@ omit =
     homeassistant/components/sensor/nederlandse_spoorwegen.py
     homeassistant/components/sensor/netdata.py
     homeassistant/components/sensor/neurio_energy.py
+    homeassistant/components/sensor/noaa-tide.py
```
Misspelled module name.

MartinHjelmare reviewed (changes requested):
The request and data parsing should be extracted to standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

jcconnell committed:
Fixing module name

jcconnell commented:
@MartinHjelmare I see that there is a python library [py_noaa](https://github.com/GClunies/py_noaa) that is already solving the request and data parsing but I noticed that they are also using the Pandas library. Pandas may be overkill for this use case or for Home Assistant in general but I don't have enough experience to say one way or another. Do you think this library would be suitable or should I write a similar library without Pandas?

MartinHjelmare commented:
Pandas is ~40 MB on disk for me. Download is ~9 MB. We already depend on numpy which is ~57 MB on disk. Those are relatively big libraries. I think it could be ok. What do you think @balloob?

balloob commented:
fine with me.

balloob committed:
Merge branch 'dev' into rc

balloob committed:
Bumped version to 0.75.0b0

superpuffin committed:
Upgrade Adafruit-DHT to 1.3.3 (#15706)

* Change to newer pip package

The package Adafruit_Python_DHT==1.3.2 was broken and would not install, breaking DHT sensor support in Home assistant. It has since been fixed in Adafruit-DHT==1.3.3.

See: https://github.com/adafruit/Adafruit_Python_DHT/issues/99

* Update requirements_all.txt

New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

* Comment out Adafruit-DHT

Adafruit_Python_DHT changed name to Adafruit-DHT, which still need pyx support breaking our CI, need to be comment out.

* Update requirements_all.txt

balloob committed:
Bumped version to 0.75.0b1

balloob committed:
Update translations

balloob committed:
Update frontend to 20180802.0

bryanyork committed:
Fix Min/Max Kelvin color temp attribute for Google (#15697)

* Fix Min/Max Kelvin color temp attribute for Google

Max Kelvin is actually Min Mireds and vice-versa. K = 1000000 / mireds

* Update test_smart_home.py

* Update test_trait.py

balloob committed:
Return True from Nest setup (#15797)

balloob committed:
Bump frontend to 20180803.0

Kane610 committed:
Only report color temp when in the correct color mode (#15791)

balloob committed:
Bumped version to 0.75.0

balloob committed:
Merge remote-tracking branch 'origin/master' into rc

balloob committed:
Merge pull request #15806 from home-assistant/rc

0.75.0

balloob committed:
Bump frontend to 20180804.0

Danielhiversen committed:
Fix rfxtrx device id matching (#15819)

* Issue #15773

Fix PT2262 devices are incorrectly matched in rfxtrx component

* style

balloob committed:
Bumped version to 0.75.1

balloob committed:
Merge pull request #15823 from home-assistant/rc

0.75.1

fabaff committed:
Upgrade voluptuous to 0.11.5 (#15830)

Cinntax committed:
Fix envisalink reconnect (#15832)

* Fix logic for handling connection lost/reconnect

* Fixed line length issue.

balloob committed:
Fix requirements

balloob committed:
Bumped version to 0.75.2

balloob committed:
Merge pull request #15856 from home-assistant/rc

0.75.2

jcconnell committed:
Use py_noaa for request and data parsing.

jcconnell committed:
Merge branch 'noaa-tides' of https://github.com/jcconnell/home-assistant into noaa-tides

jcconnell committed:
Fix style errors.

jcconnell commented:
Looks like Travis CI failed with a timeout.

MartinHjelmare commented on homeassistant/components/sensor/noaa_tides.py:
```diff
@@ -0,0 +1,128 @@
+DEFAULT_NAME = 'NOAA Tides'
+DEFAULT_TZ = 'lst_ldt'
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=60)
```
Is this the limit for the noaa api?

Please also add a `SCAN_INTERVAL` constant that sets the default scan interval of this platform.

jcconnell replied:
They don't specify a limit on their API description page. I figured 1 hour would be a safe estimate. I had removed the `SCAN_INTERVAL` that was previously present and replaced it with this. My mistake.

MartinHjelmare commented on homeassistant/components/sensor/noaa_tides.py:
```diff
@@ -0,0 +1,128 @@
+                time_zone=self._tz)
+            self.data = df_predictions.head()
+            _LOGGER.debug("Data = %s", self.data)
+            _LOGGER.info("Recent Tide data queried with start time set to %s",
```
Include this in the debug log instead.

MartinHjelmare commented on homeassistant/components/sensor/noaa_tides.py:
```diff
@@ -0,0 +1,128 @@
+        except ValueError as err:
+            _LOGGER.error("Check NOAA Tides and Currents %s", err.args)
+            self.data = None
+            raise
```
Don't raise.

MartinHjelmare commented on homeassistant/components/sensor/noaa_tides.py:
```diff
@@ -0,0 +1,128 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+API_URL = 'https://tidesandcurrents.noaa.gov/api/datagetter'
```
This isn't used anymore.

MartinHjelmare commented on homeassistant/components/sensor/noaa_tides.py:
```diff
@@ -0,0 +1,128 @@
+    def device_state_attributes(self):
+        """Return the state attributes of this device."""
+        attr = {ATTR_ATTRIBUTION: DEFAULT_ATTRIBUTION}
+        if self.data['hi_lo'][1] == 'H':
```
This will still be evaluated when state is unknown and `self.data` is `None`. So we will get an error here.

jcconnell committed:
Incorporate @MartinHjelmare.

jcconnell commented:
Thank you for the detail @MartinHjelmare! Can you help me understand why it is suggested to use the new style `string.format` in some areas but in logging it is suggested to use the other style with `%`? 

I have made the changes you've requested. I don't have a definite on the `SCAN_INTERVAL`, do you know what a good safe-estimate would be? Also, is `SCAN_INTERVAL` preferred over using a throttle? In some of my previous submissions I used Throttle and I suppose I don't understand their differences.

Again, thank you for your patience with me!

MartinHjelmare commented:
We're following the latest python convention for string formatting. Unfortunately the logging module in python still uses the old style formatting operator. Until the logging api changes we're stuck with this difference.

`SCAN_INTERVAL` should be set appropriately both according to how useful an update would be and what limit the api allows. Both these things differ between integrations so I can't make a general recommendation.

The `SCAN_INTERVAL` sets the interval the entities of a platform will call the `update` method. This is overridable by the user in the platform config.

Using a throttle on a method will limit how often that method can be called. This is most useful when there is a common client instance that talks to the api, and this client is shared between all entities of a platform. Then the throttle will limit the frequency of all communications with the api, and not just the frequency per entity. From the api perspective all requests from the same address will count towards the api limit. It doesn't matter if it's different entities that spawns the requests.

Having both a throttle and setting `SCAN_INTERVAL` can be useful if we want to set a default scan interval but have a lower limit for the scan interval that the user can't override.

jcconnell commented:
Thank you, that was very helpful! Any idea how to address the timeout error in Travis CI?

awarecan committed:
Fix downgrade hassio cannot get refresh_token issue (#15874)

* Fix downgrade hassio issue

* Update __init__.py

balloob committed:
Bumped version to 0.75.3

balloob committed:
Merge pull request #15903 from home-assistant/rc

0.75.3

jcconnell commented:
Just wanted to follow up here to see if there was something I could do to address the timeout error in TravisCI or if there are some additional changes I should make.

MartinHjelmare commented:
You could try rebasing on latest dev to see if that makes a difference.

jcconnell committed:
Merge branch 'master' of https://github.com/home-assistant/home-assistant into noaa-tides

MartinHjelmare commented:
Now the branch contains unrelated commits. So now a rebase is required. Or start over in a fresh branch, copy over changes then open a new PR.

jcconnell commented:
Sorry @MartinHjelmare! I asked for some help from someone with more Git experience but who is not familiar with the Home Assistant workflow. My apologies. I will create a new branch, reference this branch and then close it.

MartinHjelmare commented:
No worries!
